guides/hack/08-generic-types-and-functions/20-reified-generics.md

```diff
@@ -0,0 +1,187 @@
+
+```Hack
+class A {}
+class B<reify T> {}
+class C<reify Ta, Tb> {}
+
+int // enforceable
+A // enforceable
+B<int> // enforceable
+B<A> // enforceable
+C<int, string> // NOT enforceable as C's second generic is erased
+```
+
+## Creating an instance of a class with `new`
+
+Prior to reified generics, in order to create a new instance of a class without a constant class name, you'd need to pass it as `classname<T>` which is not type safe. In the runtime, classnames are strings.
+
+```Hack
+<<__ConsistentConstruct>>
+abstract class A {}
+
+class B extends A {}
+class C extends A {}
+
+function f<<<__Newable>> reify T as A>(): T {
+  return new T();
+}
+
+f<A>(); // not newable since it is abstract class
+f<B>(); // success
+f<C>(); // success
+```
+
+Notice that the reified type parameter has the attribute `<<__Newable>>`. In order for a type to be `<<__Newable>>`, the type must represent a class that's not abstract and has a consistent constructor or be a final class. Creating a new instance using the reified generics also carries across the generics given. For example,
+
+```Hack
+final class A<reify T> {}
+
+function f<<<__Newable>> reify T as A<string>>(): A<string> {
+  return new T();
+}
+
+// creates a new A<int> and since f's return type is A<string>,
+// this raises a type hint violation
+f<A<int>>();
+```
+
+## Accessing a class constant / static class property / static class method
+
+```Hack
+class C {
+  const string class_const = "hi";
+  public static string $class_propery = "hi";
+  public static function h<reify T>(): void {}
+}
+
+// Without reified generics
+function f<T as C>(classname<T> $x): void {
+  $x::class_const;
+  $x::$class_propery;
+  $x::h<int>();
+}
+
+// With reified generics
+function g<reify T as C>(): void {
+  T::class_const;
+  T::$class_propery;
+  T::h<int>();
+}
+```
+
+## Hack Arrays
+
+Hack Arrays can be used as inner type for classes since we do not need to check whether each element of the `vec` is `string`.
+Look at the limitations section for more information on when Hack Arrays cannot be used.
+
+```Hack
+class Box<reify T> {}
+
+function foo(): Box<vec<string>> {
+  return new Box<vec<int>>(); // Type hint violation
+}
+
+foo(); 
+```
+
+# Limitations
+
+* No support for subtyping (reified type parameter on classes are invariant, they cannot be co/contra-variant
+
+```Hack
+class C<reify +Ta> {} // Cannot make the generic covariant
+```
+
+* Static class methods cannot use the reified type parameter of its class
+
+```Hack
+class C<reify T> {
+  public static function f(): void {
+    return new T(); // Cannot use T
+   }
+}
+```
+
+* Hack arrays cannot be reified when used as containers with `__Enforceable` or `__Newable`
+
+```Hack
+function f<<<__Enforceable>> reify T>(): void {
+    $x is vec<int>; // Cannot use vec<int>
+    $x is T;
+}
+f<vec<int>>(); // not enforceable
+```
+
+but can be used as inner type for classes
```

Comment:
Why would `AClass<vec<int>>` be `__Enforceable`, `vec<int>` would not be? Is it because the vec<int> is not being checked, or because the type system ensures that checking it would not be required.

Reply:
How would this interact with TAny?
```HACK
// I am defined in a partial file. I take any variable and erase its type by not declaring my return type.
function get_tany(mixed $mixed){ return $mixed; }

class Box<reify T>{ public function __construct(public T $v){} }
<<__EntryPoint>>
function main(): noreturn{
  $box = new Box<vec<int>>(vec[get_t_any('not an int')]);
  invariant($box is Box<vec<int>>, "As far as I understand this will be true, even though the vec contains a string.");
  exit(0);
}